### PR TITLE
Add tasklist

### PR DESCRIPTION
**Pull request by pianocomposer321**
See #5 tasklist section.

Adds a list of tasks for each filetype so that each filetype can have more than one command tied to it (for example, one to compile, one to run, and one for tests, etc.)

Still todo:
Add global tasks for use in .yabs file.

### CODE REVIEW

**Comment by pianocomposer321** [comment]
Global tasks added!

@Shatur could you check over this to make sure this will work with what you were suggesting [here](https://github.com/pianocomposer321/yabs.nvim/issues/5#issuecomment-917711093)?

**Comment by Shatur** [comment]
Hi!

I trying the following configuration:

```lua
Yabs = require("yabs")
Yabs:setup {
    languages = {
        cpp = {
            default_task = "build&run",
            tasks = {
                build = {
                    command = "git status",
                    output = "quickfix"
                },
                run = {
                    command = "echo 1",
                    output = "quickfix"
                }
            }
        }
    }
}
```

With the following command:

```
:lua Yabs:run_task("build", "run")
```

It works, but I see the output only from the first command. Looks like the second command doesn't run. Am I missing something?

Can I also suggest to use `default_task = {"build", "run"},` instead of `default_task = "build&run",`?

File: lua/yabs/defaults/output/quickfix.lua [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
 local function on_read(lines)
     for index, line in ipairs(lines) do
         if index == #lines and line == "" then goto continue end
 
         vim.fn.setqflist({}, "a", {lines = {line}})
-        if not open and open_on_run == "auto" then
-            open = true
+        if open_on_run == "auto" then
             vim.cmd("bot copen")
             vim.cmd("wincmd p")
```
Comment (Shatur) [inline lua/yabs/defaults/output/quickfix.lua:10]: Nitpick: Use `vim.api.nvim_command` instead, it works faster. See [this](https://github.com/hrsh7th/nvim-cmp/pull/168#discussion_r703072501) comment for details.

Comment (pianocomposer321) [reply lua/yabs/defaults/output/quickfix.lua:10]: I’ll look into this. Thanks for the suggestion!

**Comment by pianocomposer321** [comment]
> With the following command:
> 
> ```
> :lua Yabs:run_default_task()
> ```
> 
> And I see the following error:
> 
> ```
> E5108: Error executing lua .../pack/plugins/start/yabs.nvim/lua/yabs/language/init.lua:59: attempt to index a nil value
> ```

> Can I also suggest to use `default_task = {"build", "run"},` instead of `default_task = "build&run",`?

If I’m not mistaken, the reason you’re getting the error is because I already implemented it in the way that you are suggesting 😂. This one’s still on me though because I didn’t provide a configuration example. I think yours is correct except for that one thing however.

Another thing, global tasks are added by adding a `tasks` table at the same level as `languages` in `yabs:setup()`. Tasks in this table will override any that are language specific (to make it useful in a `.yabs` file).

**Comment by pianocomposer321** [comment]
@Shatur a more complete configuration example would be the following:

```lua
require("yabs"):setup {
    languages = {
        lua = {
            default_task = "run",
            tasks = {
                run = {
                    command = "luafile %",
                    type = "vim"
                }
            }
        },
        c = {
            default_task = {"build", "run"},
            tasks = {
                build = {
                    command = "gcc main.c -o main",
                    output = "quickfix"
                },
                run = {
                    command = "./main",
                    output = "consolation"
                }
            }
        }
    },
    tasks = {
        build = {
            command = "echo building project...",
            output = "echo"
        },
        run = {
            command = "echo running project..",
            output = "echo"
        }
    }
}
```

**Comment by Shatur** [comment]
> If I’m not mistaken, the reason you’re getting the error is because I already implemented it in the way that you are suggesting

Oh, right, it works, thanks :) 

How to run a global task when I in `cpp` file, for example?

**Comment by pianocomposer321** [comment]
> > If I’m not mistaken, the reason you’re getting the error is because I already implemented it in the way that you are suggesting
> 
> Oh, right, it works, thanks :)
> 
> How to run a global task when I in `cpp` file, for example?

It’s the same as for non-global tasks :wink:. So `yabs:run_task(…)`.

Come to think of it, maybe there should be a way to specify which type you as the user want to run…

**Comment by Shatur** [comment]
> It’s the same as for non-global tasks wink. So yabs:run_task(…).

Got it. So global tasks will be overwrite with the current language. Sounds reasonable.

**Comment by pianocomposer321** [comment]
> > It’s the same as for non-global tasks wink. So yabs:run_task(…).
> 
> Got it. So global tasks will be overwrite with the current language. Sounds reasonable.

Actually, currently global tasks overwrite language-specific ones. I did it that way so that you can put global tasks in the `.yabs` file (which is really the only useful way to use them IMO) and they will overwrite any language-specific ones that you may have set up in your init.lua or wherever you call `yabs:setup()`.

> Can we have an API to get list of all available tasks for Telescope?

👍 I’ll work on that next.

**Comment by Shatur** [comment]
> Actually, currently global tasks overwrite language-specific ones. I did it that way so that you can put global tasks in the .yabs file (which is really the only useful way to use them IMO) and they will overwrite any language-specific ones that you may have set up in your init.lua or wherever you call yabs:setup().

Oh, I thought that I can just extend tasks (with overriding if the same task already exists) `yabs:setup()` in `.yabs` file.

**Comment by pianocomposer321** [comment]
> > Actually, currently global tasks overwrite language-specific ones. I did it that way so that you can put global tasks in the .yabs file (which is really the only useful way to use them IMO) and they will overwrite any language-specific ones that you may have set up in your init.lua or wherever you call yabs:setup().
> 
> Oh, I thought that I can just extend tasks (with overriding if the same task already exists) by using `yabs:setup()` in `.yabs` file.

You can, but if you have tasks of the same name already set up for the current filetype, the global ones you define in the `.yabs` file will never be used.

Obviously you could work around this by simply naming your global tasks something that won’t clash with your already defined language-specific ones. Maybe this is a better approach, idk. LMK what you think about this.

**Comment by Shatur** [comment]
> You can, but if you have tasks of the same name already set up for the current filetype, the global ones you define in the .yabs file will never be used.

I think this is the expected behavior :) Sounds good to me.

**Comment by pianocomposer321** [comment]
> > You can, but if you have tasks of the same name already set up for the current filetype, the global ones you define in the .yabs file will never be used.
> 
> I think this is the expected behavior :) Sounds good to me.

So, just to give an example, if you have a task for the `cpp` filetype already set up in your neovim config called `build`, you would have to call your global one in your `.yabs` file something like `build_project` instead.

I will work on this, as well as the telescope api when I get a chance.

Btw, what does the telescope api need to do? Just return a list of the current tasks (probably you would be able to specify if you want just global tasks, just the ones for the current filetype, or both)? Or is there something more complex that that needs to be?

**Comment by Shatur** [comment]
> Btw, what does the telescope api need to do? Just return a list of the current tasks (probably you would be able to specify if you want just global tasks, just the ones for the current filetype, or both)? Or is there something more complex that that needs to be?

I would go with just all available tasks for simplicity.

**Comment by pianocomposer321** [comment]
Pushed!

As the commit message states, you can get all tasks with `require("yabs"):get_tasks()`, just the ones for the current language with `require("yabs"):get_tasks_current_language()`, or only global ones with `require("yabs"):get_tasks_global()`.

@Shatur LMK if this is what you needed!

**Comment by Shatur** [comment]
Looks nice, thanks!

But maybe replace replace `require("yabs"):get_tasks_current_language()` and `require("yabs"):get_tasks_global()` with `require("yabs"):get_current_language_tasks()` and `require("yabs"):get_global_tasks()`? It makes it easy to autocomplete and makes it more human-readable.

**Comment by Shatur** [comment]
Awesome, thank you!

**Comment by pianocomposer321** [comment]
@Shatur have you tested this yet? Does it provide everything you needed? If so, I think this is read to be merged.

**Comment by Shatur** [comment]
Yes. I think it good to go.

**Pull request merged**